Pass full command name to executables

jdevera opened the pull request:
Use the ${$PREFIX}_COMMAND_NAME environment variable to pass the full command to the executable files when they are run.

The full command would be the app name, followed optionally by the group name, and finally the command name that was configured in the manifest to execute the command.

This can be used, for instance, by programs to output a meaningful usage message that refers to how the program was actually executed.

jdevera-hj committed:
Pass full command name to executables

Use the ${$PREFIX}_COMMAND_NAME environment variable to pass the full
command to the executable files when they are run.

The full command would be the app name, followed optionally by the group
name, and finally the command name that was configured in the manifest
to execute the command.

This can be used, for instance, by programs to output a meaningful usage
message that refers to how the program was actually executed.

bhou committed:
Merge branch 'main' into command_name_in_env

bhou commented on internal/context/context.go:
```diff
@@ -27,6 +27,8 @@ type LauncherContext interface {
 
 	CmdPackageDirEnvVar() string
 
+	CmdNameEnvVar() string
```
Same here, FullCmdNameEnvVar()

bhou commented on internal/frontend/default-frontend.go:
```diff
@@ -261,6 +261,15 @@ func formatExamples(examples []command.ExampleEntry) string {
 	return strings.Join(output, "\n")
 }
 
+func (self *defaultFrontend) getCommandName(group, name string) string {
```
getFullCommandName()

bhou commented on internal/frontend/default-frontend.go:
```diff
@@ -261,6 +261,15 @@ func formatExamples(examples []command.ExampleEntry) string {
 }
 
+func (self *defaultFrontend) getCommandName(group, name string) string {
+	tokens := []string{self.appCtx.AppName()}
```
This appName returned here is hard coded in the main.go file. This is used for the command launcher to get the environment variable PREFIX and determine the default configuration locations: `$HOME/.${appName}`

The code here might not cover a situation in which the user renames the binary into something else. For example, the hard-coded binary is "cl", the user could rename it to "mycl". In this case, the correct value for `CL_FULL_COMMAND_NAME` will be `mycl [group] [name]`.

jdevera replied:
Good point, I change that to use the actual executable name.

bhou commented on test/README.md:
````diff
@@ -4,7 +4,7 @@ To run all the integration test simply run the script: `integration.sh`
 ```bash
 cd command-launcher
-./test/interation.sh
+./test/integration.sh
````
Nice catch :D

bhou reviewed (changes requested):
Please rename the command to avoid ambiguity, and handle the situations that user renamed the binary itself.

bhou commented on test/packages-src/bonjour/bonjour.sh:
```diff
@@ -7,6 +7,7 @@ echo $CL_FLAG_LANGUAGE
 echo $CL_ARG_1
 
 echo "packge dir: $CL_PACKAGE_DIR"
+echo "command name: $CL_COMMAND_NAME"
```
Need to modify the bonjour.bat file as well to cover windows cases

jdevera committed:
Use actual program name for command name var

The program can be renamed after building, and the COMMAND_NAME variable
should contain the actual name, rather than the built time name.

Still resort to the build time name if the program name cannot be
obtained.

jdevera committed:
Rename variable with command to FULL_COMMAND_NAME

jdevera committed:
Add missing variable on Windows tests

jdevera committed:
Test FULL_COMMAND_NAME with renamed executable

jdevera commented:
Thanks for the review!

In my second round of changes, I:

* Renamed the variable to `${PREFIX}_FULL_COMMAND_NAME`
* Loaded it with the actual name of the executable for the launcher (still with a fallback to the configured name at build time in case of errors)
* Added the variable to the .bat file to fix windows tests
* Changed the test to run on a renamed binary, so it also checks that the actual name is used

bhou commented on test/packages-src/bonjour/bonjour.bat:
```diff
@@ -4,6 +4,7 @@ ECHO %CL_FLAG_LANGUAGE%
 ECHO %CL_ARG_1%
 
 ECHO packge dir: %CL_PACKAGE_DIR%
+ECHO command name: %CL_COMMAND_NAME%
```
This shoule be %CL_FULL_COMMAND_NAME%. It fails the integration tests.

jdevera replied:
🤦 of course, my bad. Sorry for the extra back and forth, the good news is that after a first contribution, I can see the test results and iterate before you waste time 😅 .

bhou reviewed (changes requested):
The environment variable name in .bat script is not correct, which fails the integration test

jdevera committed:
Use correct var name in Windows tests
